pydantic/_internal/_model_construction.py

```diff
@@ -241,6 +243,61 @@ def __getattr__(self, item: str) -> Any:
                 return private_attributes[item]
             raise AttributeError(item)
 
+        def mro(cls):
+            original_mro = super().mro()
+
+            if cls.__bases__ == (object,):
+                return original_mro
+
+            generic_metadata = cls.__dict__.get('__pydantic_generic_metadata__', None)
+            if not generic_metadata:
+                return original_mro
+
+            origin, args = (
+                generic_metadata['origin'],
+                generic_metadata['args'],
+            )
+            if not origin:
+                return original_mro
+
+            target_params: tuple[TypeVar] = origin.__pydantic_generic_metadata__['parameters']
+            param_dict = dict(zip(target_params, args))
+
+            # This is necessary otherwise in some case TypeVar may be same:
+            #
+            # class A(BaseModel, Generic[T]): ...
+            # class B(A[int], Generic[T]): ...
+            # class C(B[str], Generic[T]): ...
+            #
+            key = '__pydantic_inserted_mro_origins__'
```

Comment:
This `__pydantic_inserted_mro_origins__` should be added as an annotated attribute on the `BaseModel` class if we are going to use it like this. I think it would also be reasonable to add it as a field in the `__pydantic_generic_metadata__` if that makes sense, but if we are using a new attribute in `BaseModel` it should be added near this area: https://github.com/pydantic/pydantic/blob/ae7770375837020fca5c182acddae8ac66302452/pydantic/main.py#L141